ide/che-ide-gwt-app/src/main/resources/org/eclipse/che/ide/public/loader.html

```diff
@@ -12,11 +12,247 @@
+            });
+            const [_, value] = entry.split('=');
+            return decodeURIComponent(value);
+        }
+
+        function asyncGetWorkspace(workspaceId) {
+            return new Promise((resolve, reject) => {
+                const request = new XMLHttpRequest();
+                request.open("GET", '/api/workspace/' + workspaceId);
+                this.setAuthorizationHeader(request).then((xhr) => {
+                    xhr.send();
+                    xhr.onreadystatechange = () => {
+                        if (xhr.readyState !== 4) {
+                            return;
+                        }
+                        if (xhr.status !== 200) {
+                            reject(xhr.status ? xhr.statusText : "Unknown error");
+                            return;
+                        }
+                        resolve(JSON.parse(xhr.responseText));
+                    };
+                });
+            });
+        }
+
+        function setAuthorizationHeader(xhr) {
+            return new Promise((resolve, reject) => {
+                if (window._keycloak && window._keycloak.token) {
+                    window._keycloak.updateToken(5).success(() => {
+                        xhr.setRequestHeader('Authorization', 'Bearer ' + window._keycloak.token);
+                        resolve(xhr);
+                    }).error(() => {
+                        window._keycloak.login();
+                        reject('Failed to refresh token');
+                    });
+                }
+
+                resolve(xhr);
+            });
+        }
+
+        function asyncCheckServiceLink(workspace, redirectUrl) {
+            return new Promise((resolve, reject) => {
+                if (!workspace.runtime) {
+                    reject("Can't check service link: Workspace isn't RUNNING at the moment.");
+                    return;
+                }
+
+                const server = workspace.runtime.links.find(link => redirectUrl.startsWith(link));
+                if (server) {
+                    resolve(server);
+                } else {
+                    reject("Workspace doesn't have a server which matches with URL: " + redirectUrl);
+                }
+            });
+        }
+
+        function asyncGetWsToken(workspace, redirectUrl) {
+            return new Promise((resolve, reject) => {
+                if (workspace.runtime) {
+                    resolve(workspace.runtime.machineToken);
+                } else {
+                    reject("Can't get ws-token: Workspace isn't RUNNING at the moment.");
+                }
+            });
+        }
+
+        function asyncAuthenticate(redirectUrl, token) {
+            const re = new RegExp(/(https?:\/\/[^\/]+?)(?:$|\/).*/),
+                //                  \    /     \     /
+                //                  scheme    host:port
+                url = redirectUrl.replace(re, "$1" + "/authenticate?token=" + encodeURIComponent(token));
```

Comment:
Maybe it would be better to send token in `Authorization` header. @skabashnyuk @mshaposhnik WDYT?

Reply:
yes if it possible.

Reply:
Not sure if it will be easy to extract token from request header on JS side

Reply:
@mshaposhnik Could you please explain why it is needed to extract it on JS side? JwtProxy is the only one component that should extract it (as far as I understand).

Fixed workflow described in the issue

Reply:
So, no second client-side page anymore? Hm.

Reply:
Yeap, it's a bit simpler now :tada:

Reply:
+ this will require CORS routine.

Reply:
Good point!

Reply:
Ok, we proved that works - fine!